typing/typecore.ml

```diff
@@ -3257,7 +3257,7 @@ and type_expect_
     in
     if is_format then
       let format_parsetree =
-        { (type_format loc str env) with pexp_loc = sexp.pexp_loc }  in
+        { (type_format loc str delim env) with pexp_loc = sexp.pexp_loc }  in
```

Comment:
Instead of passing `delim`, you could try passing `cst`, the constant string. Then maybe it works to do

```ocaml
      mk_constr "Format" [ mk_fmt fmt; mk_cst source_format_string_cst ]
```

on line 4982.

Reply:
in our case, every string must be created with the delimiter if we're going to preserve the escaping further down the pipeline.

Perhaps it's not such a good idea to submit this upstream and keep it as a local patch instead.

Reply:
I am not sure I follow. What does "preserving the ECMAscript escaping convention" mean? (I this is a matter of lexical conventions, why are they not handled during parsing?)

Reply:
In OCaml, as you know, strings are just byte sequences, so if you type, e.g. `"你"` in a UTF-8 encoded source file, the OCaml lexer will give us `"\xe4\xbd\xa0"`.

However, users of Melange may instead want to emit `"你"` verbatim. In such cases, we reserve e.g. the `js` delimiter (`{js|你|js}`) and parse the string back into unicode, that we then emit in the resulting JavaScript artifact.

[This article](https://mathiasbynens.be/notes/javascript-escapes) goes into some more detail on the JavaScript escaping that I mentioned too.

Reply:
Two remarks.

In terms of semantics of "upstream OCaml" the change is a bit dubious, as quoted string literals are supposed to have the exact same semantics as non-quoted string literals. We support quoted string *extensions* `{%foo bar|...|bar}` as shorthand syntax for extensions `[%foo {bar|...|}]`, to which code consumers can give arbitrary semantics. Such extension nodes would be a fine way to present, at the language levels, string literals that follow a different lexical convention. (In fact I see that Melange supports a `%raw` extension to write JS code directly, so `{js|...|js}` is a shorthand for `{%raw|"..."|}`.)

I tried to understand what the point of the `{js|...|js}` markers is in Melange, but your explanations remained obscure to me (I don't know what "parse the string back into unicode" means). Initially I thought that you wanted to produce utf16-encoded strings instead of utf8-encoded strings. In fact my current understanding from reading https://github.com/melange-re/melange/blob/da7b44e176f9212dafc0e199866bc1b1e6977ad1/jscomp/core/js_dump.ml#L644-L654 and https://github.com/melange-re/melange/blob/da7b44e176f9212dafc0e199866bc1b1e6977ad1/jscomp/core/js_dump_string.ml#L54 is as follows:
1. This is about the readability of the Javascript code produced by the Melange compiler, not the semantics of the OCaml program.
2. "unicode" literals, the `{js|...|js}` strings in input syntax, are compiled into string literals containing their raw content in the Javascript file (... so they better use the Javascript escaping convention).
3. "non-unicode" literals are compiled into an equivalent series of escape codes for all non-ascii characters

The natural thing would be for Melange to convert all OCaml string literals into the corresponding Javascript string literals, so that escape sequences are compiled to escape sequences and "raw" unicode characters are compiled to "raw" unicode characters. This is however not possible with the current OCaml frontend, which interprets string literals at lexing time and does not preserve the distinction between say `"a"` and `"\u{61}"`.

I do wonder still if it wouldn't be simpler to pick a strategy for normal string literals that prints more characters directly and escapes less often. Surely there are existing unicode libraries that propose criteria for "this unicode grapheme could be output raw, and will be readable in most contexts" vs "you better escape this one", and they could be used to compile all OCaml string literals into Javascript string literals, with `你` non-escaped?